enhancements/cluster-validation-tool.md

```diff
@@ -0,0 +1,521 @@
+be able to validate the configuration against the known working reference configuration to identify
+potential issues before they impact end users, service level agreements, or cluster uptime.
+
+This enhancement describes a tool capable of doing an "intelligent" diff between a reference
+configuration and a set of CRs representative of a deployed "production" cluster. These CRs may
+derive from many potential sources such as being pulled from a live cluster, read from a git
+repository, or extracted from a support archive. The reference configuration is sufficiently
+annotated to describe expected user variations versus required content.
+
+Existing tools meet some of this need but fall short of the goals
+- kubectl/oc diff: This tool allows comparison of a live cluster against a known configuration.
+  There are two shortcomings we need to address:
+  - Ability to handle expected user variation vs required content.
+  - Consumption of an offline representation of the cluster configuration.
+- `<get cr> | <key sorting> | diff` : There are various ways of chaining together existing tools
+  to obtain, correlate, and compare/diff two YAML objects. These methods fall short in 1) Ability
+  to handle expected user variation vs required content.
+
+The proposal describes both the representational format of the reference configuration as well as
+the functionality of the tool which will perform the comparison.
+
+### User Stories
+
+As an end customer I want to be able to compare the configuration of my clusters against the
+validated reference configuration. I want the tool to report significant drift and to suppress
+"diffs" which are expected user variations and runtime variable fields (eg status/metadata). I want
+the tool to report drift in deployed CRs as well as required CRs which are missing. I want to be
+able to run the tool at various points in the deployment lifecycle including against:
+* my live production clusters
+* source (eg generic/templated) CRs in my planning labs
+* reference CRs provided by partner
+
+As a support engineer helping to triage/debug issues with a customer or partner I want to be able to
+compare the configuration of a cluster as captured in a support archive against a validated
+reference configuration.
+
+As a provider of reference configurations I want to be able to provide a new or updated validated
+reference configuration without the need to rebuild the validation tool.
+
+### Goals
+
+The following goals apply to the validation (diff) tool:
+1. Data driven. New and updated reference configurations do not require a new release of the tool.
+1. Can be used with new and updated reference configurations
+1. Can consume input configuration CRs from live cluster
+1. Can consume input configuration CRs from a support archive
+1. Can consume input from local filesystem (files, directories, etc)
+1. Will suppress diffs (not report as a drift) for runtime variable fields (status, managed
```

Comment:
Should the list of runtime fields be made configurable / extensible?

Reply:
First, thanks for the really good review comments!

Initially I think we should follow the behavior of `kubectl/oc diff` in terms of fields skipped during comparison. If needed, yes we can extend the list and/or make it customizable (eg as a phase 2 enhancement).